### PR TITLE
Fix Doctests for 0.5-dev

### PR DESCRIPTION
**Pull request by omus**
Updates the doctests to be compliant with Julia 0.5-dev. This PR is based off of the work done in #15753.

Doctests currently have 14 remaining failures that can be categorized as:
- ~~Regression for display of tuples of arrays (#17087)~~
- ~~Regression for display of `macroexpand` (#17090)~~
- ~~`dump` function no longer prints trailing newline (#17202 / [julia-dev forum post](https://groups.google.com/forum/#!topic/julia-dev/uHpfvBkClTc))~~
- Regression for `jl_` printing (#17107)

Requires that JuliaLang/JuliaDoc#23 be merged. (DONE)

### CODE REVIEW

**Comment by vchuravy** [comment]
Thanks for picking this up

File: base/docs/helpdb/Base.jl [hunk 1]
```diff
@@ -5643,7 +5643,8 @@ julia> deleteat!([6, 5, 4, 3, 2, 1], 1:2:5)
 
 julia> deleteat!([6, 5, 4, 3, 2, 1], (2, 2))
 ERROR: ArgumentError: indices must be unique and sorted
- in deleteat! at array.jl:543
+ in deleteat!(::Array{Int64,1}, ::Tuple{Int64,Int64}) at ./array.jl:571
+ in eval(::Module, ::Any) at ./boot.jl:231...
```
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: same comment as before, doesn't the doctest system already do some pruning of backtraces and wouldn't this be exactly the kind of thing to add to it?

Comment (omus) [inline base/docs/helpdb/Base.jl]: The doctest system was doing some pruning before but it seemed to only work for entire lines. The ellipsis works much better for us as sometimes we append extra information afterwards such as `(repeats 2 times)`:
```
ERROR: ArgumentError: indices must be unique and sorted
 in deleteat!(::Array{Int64,1}, ::Tuple{Int64,Int64}) at ./array.jl:571
 in eval(::Module, ::Any) at ./boot.jl:231 (repeats 2 times)
 in _start() at ./client.jl:359
```

Comment (tkelman) [inline base/docs/helpdb/Base.jl]: makes sense, but the `eval(::Module, ::Any)` part isn't really useful information - would it work if we put the ellipsis at the end of the previous line?

Comment (omus) [inline base/docs/helpdb/Base.jl]: I am ok with doing that change. I didn't do that as the original stack traces included `eval(::Module, ::Any)`. 

@tkelman just to confirm: You want me to removing the `eval(::Module, ::Any)` from all (almost all) stack traces in the manual?

Comment (tkelman) [inline base/docs/helpdb/Base.jl]: What do you mean by "original stack traces" ? Like what you get at the REPL or from running this code in a script (they'll look slightly different)? I think we already have the docsystem cutting out the pieces that appear in every single backtrace, 0.5 just adds a new such line. Since it's mostly noise I think it would be better not to add it. But would like to hear other opinions on the matter too.

Comment (omus) [inline base/docs/helpdb/Base.jl]: > What do you mean by "original stack traces" ?
I meant what existed in the manual prior to this PR

Comment (tkelman) [inline base/docs/helpdb/Base.jl]: Before this PR I'm seeing a handful, but quite a bit fewer than this PR is adding:
```
$ git grep -n 'in eval'
base/test.jl:914: in eval(::Module, ::Any) at ./boot.jl:226
doc/manual/mathematical-operations.rst:388:     in eval at ./boot.jl:263
doc/manual/mathematical-operations.rst:396:     in eval at ./boot.jl:263
doc/manual/mathematical-operations.rst:401:     in eval at ./boot.jl:263
doc/manual/mathematical-operations.rst:416:     in eval at ./boot.jl:263
doc/stdlib/test.rst:283:        in eval(::Module, ::Any) at ./boot.jl:226
test/core.jl:336:@test_throws ErrorException eval(:(abstract Qux2_{T} <: Sup2b_{Qux2_{Int}, T})) # wrapped in eval to avoid #16793
```

Comment (omus) [inline base/docs/helpdb/Base.jl]: I stand corrected. It turns out that ce62f34 added in the eval line which I've been ignoring in my diffs. I'll remove those lines and probably most of the ellipsis as well.

**Comment by omus** [comment]
Things are in a much better state now. I think this can be merged before the linked issues are dealt with as we can always fix the related doctests later.

**Comment by omus** [comment]
Just waiting on #17202. I've left the doctests for #17107 broken as that issue requires that a section of the manual be re-written.

**Comment by omus** [comment]
Rebased. Fixed some line numbers that were changed. One new doctest failure has appeared with "verbose_fussy_sqrt" reporting a line number of 0 in a stack trace #17317

**Comment by kshyatt** [comment]
Thanks for your patience and all your hard work on this, @omus.

**Comment by omus** [comment]
No problem. I found the manual invaluable when learning Julia. Keeping things up to date is the least I can do.

**Comment by omus** [comment]
Rebased again since #17202 has been merged. Doctests uncovered #17338 which make up 8/11 currently failing doctests. I additionally changed one of the doctests to be skipped (8683558) as it tended to give allot of false positives.

**Comment by omus** [comment]
Another rebase since #17338 was fixed. Only three remaining failed doctests. Two require the documentation to be updated (#17107) and the other should be addressed by #17324.

File: base/test.jl [hunk 2]
```diff
@@ -893,24 +893,19 @@ julia> typeof(f(1,2,3))
 Int64
 
 julia> @code_warntype f(1,2,3)
-Variables:
```
Comment (tkelman) [inline base/test.jl]: is Variables output no longer being included in `@code_warntype` intentional?

Comment (omus) [inline base/test.jl]: It was intentional. I can add it back if it is preferred. The blank line was being an issue and I felt that the variable listing didn't add anything to the example.

File: doc/manual/control-flow.rst [hunk 3]
```diff
@@ -126,7 +126,7 @@ This means that new variables defined inside the ``ìf`` clauses can
 be used after the ``if`` block, even if they weren't defined before.
 So, we could have defined the ``test`` function above as
 
-.. doctest::
+.. doctest:: leaky
```
Comment (tkelman) [inline doc/manual/control-flow.rst]: where are these `leaky`, `bad-leaky`, etc documented?

File: base/test.jl (hunk 2 above)
Comment (tkelman) [inline base/test.jl]: oh right I see the `...` now, I guess it's alright to leave it out

File: doc/manual/documentation.rst [hunk 4]
```diff
@@ -30,7 +30,7 @@ macros and pass them to the ``@doc`` macro just as well.
 
 Here is a more complex example, still using Markdown:
 
-.. doctest::
+.. code-block:: julia
```
Comment (tkelman) [inline doc/manual/documentation.rst]: I guess this wasn't outputting anything?

File: doc/manual/metaprogramming.rst [hunk 5]
```diff
@@ -415,11 +418,16 @@ Basics
...
+.. testcode::
 
-    julia> macro sayhello()
-               return :( println("Hello, world!") )
-           end
+    macro sayhello()
+        return :( println("Hello, world!") )
+    end
+
+.. testoutput::
```
Comment (tkelman) [inline doc/manual/metaprogramming.rst]: so this runs as a doctest, checks the output matches but doesn't include the output in the rendered docs?

File: doc/manual/metaprogramming.rst [hunk 6]
```diff
@@ -620,11 +629,14 @@ Compare:
     Expr
       head: Symbol string
       args: Array{Any}((5,))
-        1: String "a ("
+        1: String
+          data: Array{UInt8}((3,)) UInt8[0x61,0x20,0x28]
```
Comment (tkelman) [inline doc/manual/metaprogramming.rst]: hm I guess this was a little nicer in the old format? is this a change in `dump` ?

File: doc/manual/methods.rst [hunk 7]
```diff
@@ -226,148 +235,29 @@ values is perhaps the single most powerful and central feature of the
 Julia language. Core operations typically have dozens of methods:
 
 .. doctest::
+   :options: +SKIP
```
Comment (tkelman) [inline doc/manual/methods.rst]: I might take this out right before release and see how annoying it gets while backporting fixes

File: doc/manual/methods.rst [hunk 8]
```diff
@@ -427,9 +312,9 @@ the intersection case:
     julia> g(2.0, 3.0)
     10.0
 
-To suppress Julia's warning, the disambiguating method must be defined
-first, since otherwise the ambiguity exists, if transiently, until the
-more specific method is defined.
+It is recommended that the disambiguating method be defined first,
+since otherwise the ambiguity exists, if transiently, until the more
```
Comment (tkelman) [inline doc/manual/methods.rst]: maybe rephrase "if transiently" to "albeit transiently" ?

File: doc/manual/metaprogramming.rst (hunk 6 above)
Comment (tkelman) [inline doc/manual/metaprogramming.rst]: bisected to 3405d263194 so I guess it was intentional - the new output isn't quite as nice, but I guess it says "`dump` should not be overloaded" so I guess we're stuck with it

File: doc/manual/metaprogramming.rst (hunk 5 above)
Comment (omus) [inline doc/manual/metaprogramming.rst]: Correct. The original doctest was failing since creating the macro now was outputting:
```
@sayhello (macro with 1 method)
```
Using `testcode` and `testoutput` allowed me to keep the example tested while keeping the documentation the same. Note we did lose the `julia>` prompt.

File: doc/manual/documentation.rst (hunk 4 above)
Comment (omus) [inline doc/manual/documentation.rst]: I changed this to a code block as Julia interprets `function bar(x, y) ...` as a function call to `bar` which raises an exception.

File: doc/manual/metaprogramming.rst (hunk 6 above)
Comment (omus) [inline doc/manual/metaprogramming.rst]: I agree the output isn't as nice. Would it be worth opening an issue to confirm the changes to String output was intentional?

File: doc/manual/methods.rst (hunk 7 above)
Comment (omus) [inline doc/manual/methods.rst]: Fair enough. If we do want to require doctests to pass for PRs I think this particular doctest is going to be the main issue.

**Comment by omus** [comment]
Thanks for merging this @tkelman

File: doc/manual/metaprogramming.rst (hunk 6 above)
Comment (tkelman) [inline doc/manual/metaprogramming.rst]: might get closed with a "yes," maybe a comment on that dump pr to start with?

File: doc/manual/documentation.rst (hunk 4 above)
Comment (tkelman) [inline doc/manual/documentation.rst]: strange, I think it's been passing on release-0.4?

File: doc/manual/metaprogramming.rst (hunk 6 above)
Comment (omus) [inline doc/manual/metaprogramming.rst]: I decided to just make an issue as GitHub comments are hard to search for #17392
